src/mmda/predictors/heuristic_predictors/dictionary_word_predictor.py

```diff
@@ -1,27 +1,76 @@
 """
-DictionaryWordPredictor -- Reads rows of text into dehyphenated words.
+DictionaryWordPredictor -- Reads `tokens` and converts them into whole `words`.
+
+Let's consider 4 consecutive rows:
+
+    This is a few-shot learn-
+    ing paper. We try few-
+    shot techniques. These
+    methods are useful.
+
+This predictor tries to:
+    1. merge tokens ["learn", "-", "ing"] into "learning"
+    2. merge tokens ["few", "-", "shot"] into "few-shot"
+    3. keep tokens ["These", "methods"] separate
+    4. keep tokens ["useful", "."] separate
+
+This technique requires 2 passes through the data:
+
+    1. Build a dictionary of valid words, e.g. if "learning" was ever used, then we can merge
+       ["learn", "-", "ing"].
+
+    2. Classify every pair of tokens as belonging to the *same* or *different* words.
+
+@kylel, @rauthur
 
-@rauthur
 """
 
-import string
-from typing import Optional, Set, List
+import os
+from typing import Optional, Set, List, Tuple, Iterable, Dict
 
+from collections import defaultdict
+from mmda.parsers import PDFPlumberParser
 from mmda.predictors.base_predictors.base_predictor import BasePredictor
-from mmda.types.metadata import Metadata
-from mmda.types.annotation import Annotation, Span, SpanGroup
-from mmda.types.document import Document
+from mmda.predictors import WhitespacePredictor
+from mmda.types import Metadata, Document, Span, SpanGroup
 from mmda.types.names import Rows, Tokens
 
 
-class DictionaryWordPredictor(BasePredictor):
+class Dictionary:
+    def __init__(self, words: Iterable[str], punct: str):
+        self.words = set()
+        self.punct = punct
+        for word in words:
+            self.add(word)
+
+    def add(self, text: str) -> None:
+        self.words.add(self.strip_punct(text=text.strip().lower()))
+
+    def is_in(self, text: str, strip_punct=True) -> bool:
+        if strip_punct:
+            return self.strip_punct(text=text.strip().lower()) in self.words
+        else:
+            return text.strip().lower() in self.words
+
+    def strip_punct(self, text: str) -> str:
+        start = 0
+        while text[start] in self.punct and start < len(text):
+            start += 1
+        end = len(text) - 1
+        while text[end] in self.punct and end > 0:
+            end -= 1
+        return text[start: end + 1]
```

Comment:
```suggestion
        assert start >= end
        return text[start: end + 1]
```

Reply:
why is this necessary?

Reply:
It is not necessary disregard, I thought maybe it is a good check